enhancements/network/ovn-secondary-networks.md

```diff
@@ -0,0 +1,240 @@
+---
+title: ovn-secondary-networks
+authors:
+  - "@maiqueb"
+reviewers:
+  - "@dcbw"
+  - "@dougbtv"
+  - "@squeed"
+  - "@trozet"
+  - "@abhat"
+  - "@alexanderConstantinescu"
+  - TBD
+approvers:
+  - TBD
+creation-date: 2021-06-15
+last-updated: yyyy-mm-dd
+status: provisional
+---
+
+# OVN secondary networks
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+This enhacement proposal studies and describes how to support OVN as the
+networking provider for a Multus secondary interface.
+
+## Motivation
+The networking plugins that can be used as secondary interfaces are currently
+quite simple, and are unable to provide advanced networking requirements.
+
+Furthermore, virtualization users are used to being able to define isolated
+tenant networks that form a single L2 domain. OVN excels at this, by creating
+simple L2 overlays without the need for a network administrator.
+
+### Goals
+- create flat layer 2 overlays
+- connect a newly created pod to these overlay networks
+- these overlays can optionally be a subnet
+- IPv6 support is welcome (not an immediate priority)
+- optional connection to the external networks
+
+### Non-Goals
+- network policy concerns. There isn't a requirement for having network
+  policies for the secondary networks. There is an
+  [enhancement proposal](enhancements/network/multi-networkpolicy.md) about
+  this subject, not scoped around OVN.
+
+## Proposal
+
+### User Stories
+
+#### Isolated tenant network
+A customer can specify a layer 2 overlay network with a subnet via a
+`network-attachment-definition`, which is provisioned without the intervention
+of a system administrator. You can refer to the
+[network attachment definition section](#net-attach-def-samples) for examples.
+
+Afterwards, the users can then request their pods - and / VMs - to connect to
+those overlays as they would for any other secondary network.
+
+### Implementation Details/Notes/Constraints
+A new CRD will be introduced, named `OVNSecondaryNetwork`. When this CRD is
+provisioned, a `NetworkAttachmentDefinition` object will be rendered.
+
+The new CRD will look like:
+
+```golang
+type OVNSecondaryNetwork struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec OVNSecondaryNetworkSpec `json:"spec,omitempty"`
+}
+
+type OVNSecondaryNetworkSpec struct {
+    // Subnet is a RFC 4632/4291-style string that represents an IP address and prefix length in CIDR notation
+    Subnet                  string `json:"subnet"`
```

Comment:
How are subnets validated? Do we prevent overlapping subnets with other things in the cluster? (other secondary networks, node CIDRs for the primary network, ClusterCIDR, serviceCIDR, etc)

Reply:
Two things:
1 - Can you explain what this field does? Does it provide IPAM, or is it just for routing? Does it cause a gateway to be created (at the .1)? Too much magic in this field...
2 - This should always be an array for dual-stack (or just disjoint ranges).

Reply:
> Two things:
> 1 - Can you explain what this field does? Does it provide IPAM, or is it just for routing? Does it cause a gateway to be created (at the .1)? Too much magic in this field...

I'll explain better.

> 2 - This should always be an array for dual-stack (or just disjoint ranges).

I am proposing a very simple model here; one IP per interface, and 1 subnet per logical switch. If the user desires dual stack, the option would be to create 2 `OverlayNetwork`s, each with an IP version.

Is this model too narrow minded ?

Reply:
Starting simple is fine. However, let's just make this dual-stack from the get-go.

A convention is to have arrays for all IPs, and reject as invalid if there is more than one ip / cidr per family.